funnel/models/membership.py

```diff
@@ -0,0 +1,436 @@
+        roles.update(self.parent.roles_for(actor, anchors))
+        return roles
+
+
+# Add active membership relationships to Profile and User
+
+Profile.active_admin_memberships = db.relationship(
+    ProfileAdminMembership,
+    lazy='dynamic',
+    primaryjoin=db.and_(
+        ProfileAdminMembership.profile_id == Profile.id, ProfileAdminMembership.active
+    ),
+)
+
+Profile.active_owner_memberships = db.relationship(
+    ProfileAdminMembership,
+    lazy='dynamic',
+    primaryjoin=db.and_(
+        ProfileAdminMembership.profile_id == Profile.id,
+        ProfileAdminMembership.active,
+        ProfileAdminMembership.is_owner.is_(True),
+    ),
+)
+
+User.active_profile_admin_memberships = db.relationship(
+    ProfileAdminMembership,
+    lazy='dynamic',
+    primaryjoin=db.and_(
+        ProfileAdminMembership.user_id == User.id, ProfileAdminMembership.active
+    ),
+)
+
+User.active_profile_owner_memberships = db.relationship(
+    ProfileAdminMembership,
+    lazy='dynamic',
+    primaryjoin=db.and_(
+        ProfileAdminMembership.user_id == User.id,
+        ProfileAdminMembership.active,
+        ProfileAdminMembership.is_owner.is_(True),
+    ),
+)
+
+User.profiles_owned = association_proxy('active_profile_owner_memberships', 'profile')
+
+
+class ProjectCrewMembership(ImmutableMembershipMixin, db.Model):
+    """
+    Users can be crew members of projects, with specified access rights.
+    """
+
+    __tablename__ = 'project_crew_membership'
+
+    # List of is_role columns in this model
+    __role_columns__ = ('is_editor', 'is_concierge', 'is_usher')
+    __parent_column__ = 'project_id'
+
+    __roles__ = {
+        'all': {
+            'read': {'user_details', 'is_editor', 'is_concierge', 'is_usher', 'project'}
+        },
+        'profile_admin': {
+            'read': {'edit_url', 'delete_url', 'project'},
+            'call': {'url_for'},
+        },
+        # 'profile_owner': {
+        #     'read': {
+        #         'user_details',
+        #         'is_editor',
+        #         'is_concierge',
+        #         'is_usher',
+        #     }
+        # },
+    }
+
+    project_id = immutable(
+        db.Column(None, db.ForeignKey('project.id', ondelete='CASCADE'), nullable=False)
+    )
+    project = immutable(
+        db.relationship(
+            Project,
+            backref=db.backref(
+                'crew_memberships',
+                lazy='dynamic',
+                cascade='all, delete-orphan',
+                passive_deletes=True,
+            ),
+        )
+    )
+    parent = immutable(db.synonym('project'))
+
+    user_id = immutable(
+        db.Column(
+            None,
+            db.ForeignKey('user.id', ondelete='CASCADE'),
+            nullable=False,
+            index=True,
+        )
+    )
+    user = immutable(
+        db.relationship(
+            User,
+            foreign_keys=[user_id],
+            backref=db.backref(
+                'profile_crew_memberships',
```

Comment:
Now I remember why we kept the `user` relationship in the subclass instead of the mixin: for the backrefs. Still, this needs a `@with_roles(..., grants={'subject'})`.